site/topics/genetic-programming/symbolic-regression-gp.rst

```diff
@@ -116,9 +116,175 @@ DEAP
+      - 2.63416737e+02
+
+    * - ...
+      - ...
+      - ...
+
+
+* For this example, the data is provided in CSV files
+* The data is formatted such that the last column is always the dependant variable
+
+    * Thus, all proceeding columns are independent variables
+
+
+* This data was generated by some phenomenon
+* The goal is to find a model that effectively describes the relationships between the data
+* Loading this data can be done as follows
+
+.. literalinclude:: /../src/gp_symbolic_regression.py
+    :language: python
+    :lineno-match:
+    :start-after: # [begin-data]
+    :end-before: # [end-data]
+
+
+* The above code loads the data specified in some constants
+* The independent variables will be stored in a two dimensional list
+
+    * Rows are observations
+    * Columns are different independent variables
+
+
+* The dependent variable is stored in a one dimensional list
+
+    * Each value corresponds to the independent variables at the same index
+
+
+* Following the example data in the above table, the values would be stored as follows
+
+.. code-block:: python
+
+    independent_variables = [[-3.16833493e+01, 3.63253949e+01],
+                            [-3.59771849e+01, 1.63633781e+01],
+                            [4.35235243e+01, -1.54189209e+01],
+                            [4.19705508e+01, 1.25491733e+01],
+                            ... ]
+
+    dependant_variable = [-5.75550374e+02, -2.94238014e+02, -3.35451435e+02, 2.63416737e+02, ...]
+
+
+
+Evaluation
+==========
+
+* A common mechanism for evaluating genetic programming performing symbolic regression is *mean squared error*
+
+    * :math:`\frac{1}{n}\sum_{i}^{n}(y_{i} - \hat{y_{i}})^{2}`
+
+
+* A mean squared error of zero is perfect
+
+* Any metric could be used, but mean squared error is very popular
+* One of the reasons it is used is because it *punishes* bigger errors more
+
+    * If the difference between the observed and estimated value is 1, the squared error is 1
+    * If the difference is 2, the error is 4
+    * If it's 3, the error is 9
+    * ...
+    * In other words, it's much worse to be very wrong than it is to be a little wrong
+
+
+.. literalinclude:: /../src/gp_symbolic_regression.py
+    :language: python
+    :lineno-match:
+    :pyobject: mean_squared_error
+
+
+
 Language
 ========
 
+* With symbolic regression, genetic programming will be searching the space of mathematical expressions
+* This includes both the operators and operands
+* The language is the set of operators and operands that the genetic programming system can use in the search
+* The trouble is, it's not always clear which operators and operands should be included
+* Thus, the design of the language may need to be tuned for evolution like other hyperparameters
+
+* Fortunately, with symbolic regression, there is a common set to start with
+
+    * Typical arithmatic operators
+    * Maybe some trigonometry functions
+    * Maybe natural log and Euler's number
```

Comment:
maybe _Maybe logarithm and exponentials_

Reply:
👍